### DISCUSSION TITLE
way to run the optimiser offline?

### DISCUSSION THREAD

**Discussion by racitup** [SmileyChris/easy-thumbnails#475] [Q&A]
I'm deploying to heroku except it seems the optimiser runs on a request basis, so I get errors like the following because once the app 'slug' is built, heroku doesn't let the application write to the filesystem:

> 2017-10-03T22:18:25.695191+00:00 app[web.1]: [2017-10-03 23:18:25,694 easy_thumbnails.optimize] ERROR: Command '/usr/local/bin/optipng /tmp/tmp7y4Txw' returned non-zero exit status 127

Is there a way to run it offline?

**Comment by jrief** [comment]
as long as you can run the thumbnailer offline, then the optimizer it is invoked automatically.
The problem is, that only during template rendering, easy-thumbnail "sees", which images shall be generated. And since the thumb sizes come from the template context, nobody can tell easy-thumbnail, which thumbs have to be generated.

This IMO can only be done offline, if you prefetch all possible templates/pages and afterwards sync your media folder to wherever it should be served from.

**Comment by racitup** [comment]
Hmm, yes that's what I thought.
I have a feeling that the optimizer saves the optimised files in a temporary directory. I have a lot of thumbnails generated in the media folder, but it seems whenever I deploy my app the optimizer is run again. Is that right?
